SSA/Projects/InstCombine/Test.lean

```diff
@@ -383,3 +383,176 @@ def three_inst_macro_noreduc_proof (w : Nat) :
+  simp_alive_ssa
+  apply two_inst_concrete_stmt
+
+def two_inst_concrete_macro_noreduc_proof  :
+  two_inst_concrete_macro_noreduce ⊑ two_inst_concrete_macro_noreduce := by
+  unfold two_inst_concrete_macro_noreduce
+  simp_alive_meta
+  simp_alive_ssa
+  apply two_inst_concrete_stmt
+
+def three_inst_concrete_macro :=
+  [alive_icom ()|{
+  ^bb0(%arg0: i1):
+    %0 = "llvm.not" (%arg0) : (i1) -> (i1)
+    %1 = "llvm.not" (%0) : (i1) -> (i1)
+    %2 = "llvm.not" (%1) : (i1) -> (i1)
+    "llvm.return" (%2) : (i1) -> ()
+  }]
+
+set_option ssa.alive_icom_reduce false in
+def three_inst_concrete_macro_noreduce :=
+  [alive_icom ()|{
+  ^bb0(%arg0: i1):
+    %0 = "llvm.not" (%arg0) : (i1) -> (i1)
+    %1 = "llvm.not" (%0) : (i1) -> (i1)
+    %2 = "llvm.not" (%1) : (i1) -> (i1)
+    "llvm.return" (%2) : (i1) -> ()
+  }]
+
+def three_inst_concrete_com :
+  Com InstCombine.LLVM [InstCombine.Ty.bitvec 1] .pure (InstCombine.Ty.bitvec 1) :=
+  Com.lete (not 1 0) <|
+  Com.lete (not 1 0) <|
+  Com.lete (not 1 0) <|
+  Com.ret ⟨0, by simp [Ctxt.snoc]⟩
+
+def three_inst_concrete_stmt (e : LLVM.IntW 1) :
+  @BitVec.Refinement (BitVec 1) (LLVM.not (LLVM.not (LLVM.not e)))
+    (LLVM.not (LLVM.not (LLVM.not e))) := by simp
+
+def three_inst_concrete_com_proof :
+  three_inst_concrete_com ⊑ three_inst_concrete_com := by
+  unfold three_inst_concrete_com
+  simp only [simp_llvm_wrap]
+  simp_alive_meta
+  simp_alive_ssa
+  apply three_inst_concrete_stmt
+
+def three_inst_concrete_macro_proof :
+  three_inst_concrete_macro ⊑ three_inst_concrete_macro := by
+  unfold three_inst_concrete_macro
+  simp_alive_meta
+  simp_alive_ssa
+  apply three_inst_concrete_stmt
+
+def three_inst_concrete_macro_noreduc_proof :
+  three_inst_concrete_macro_noreduce ⊑ three_inst_concrete_macro_noreduce := by
+  unfold three_inst_concrete_macro_noreduce
+  simp_alive_meta
+  -- How can I avoid the `simp! only`and use a plain `simp only`?
+  simp! only [ConcreteOrMVar.instantiate_list]
+  simp_alive_ssa
+  apply three_inst_concrete_stmt
+
+set_option ssa.alive_icom_reduce false in
+def three_inst_concrete_macro_noreduce_ne (w : Nat) :=
+  [alive_icom (w)|{
+  ^bb0(%arg0: _, %arg1: _):
+    %0 = "llvm.icmp.ne" (%arg0, %arg1) : (_, _) -> (i1)
+    %1 = "llvm.icmp.ne" (%arg0, %arg1) : (_, _) -> (i1)
+    "llvm.return" (%1) : (i1) -> ()
+  }]
+
+def three_inst_concrete_stmt_ne {a b : LLVM.IntW w} :
+  @BitVec.Refinement (BitVec 1) (LLVM.icmp LLVM.IntPredicate.ne a b)
+    (LLVM.icmp LLVM.IntPredicate.ne a b) := by simp
+
+def three_inst_concrete_macro_noreduc_proof_ne :
+  three_inst_concrete_macro_noreduce_ne x ⊑ three_inst_concrete_macro_noreduce_ne x := by
+  unfold three_inst_concrete_macro_noreduce_ne
+  simp_alive_meta
+  -- There remains a 'toMap' in the following expression:
+  -- (↑↑(Ctxt.Var.last ∅ (MTy.bitvec (ConcreteOrMVar.mvar 0)))).toMap
+  -- I am unsure why
+  simp only [Ctxt.Var.toSnoc_toMap']
+  simp_alive_meta
+  simp_alive_ssa
+  intros -- Why do I need an intros?
```

Comment:
Anyone knows why I need this `intros`?

Reply:
This got resolved by flipping a and b, it seems for this apply to work the variable order matters. In retrospective, this is not surprising.